### PR TITLE
Refs #6896 Added parser-generator roundtrip test.

### PR DESCRIPTION
**Pull request by LuisGP**
Small fixes were needed to pass it ;).

### CODE REVIEW

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/xtypes-manual-linux/41/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/xtypes-manual-linux/44/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/xtypes-manual-linux/45/)

File: include/xtypes/Module.hpp [hunk 1]
```diff
@@ -58,7 +58,7 @@ class Module
 
 
     bool has_submodule(
-            const std::string& submodule)
+            const std::string& submodule) const
```
Comment (lemunozm) [inline include/xtypes/Module.hpp:61]: :D

File: include/xtypes/Module.hpp [hunk 2]
```diff
@@ -278,10 +282,17 @@ class Module
...
+    {
+        return std::find(from_enum_.begin(), from_enum_.end(), name) != from_enum_.end();
+    }
+
     bool create_constant(
             const std::string& name,
             const DynamicData& value,
-            bool replace = false)
+            bool replace = false,
+            bool from_enumeration = false)
```
Comment (lemunozm) [inline include/xtypes/Module.hpp:282]: Maybe I don't understand it well, but, don't you catch the enumeration type from the DynamicData kind?

File: include/xtypes/Module.hpp [hunk 3]
```diff
@@ -335,6 +350,26 @@ class Module
...
     }
 
+    const EnumerationType<uint32_t>& enum_32(
+            const std::string& name) const
+    {
+        // Solve scope
+        PairModuleSymbol module = resolve_scope(name);
+        if (module.first == nullptr)
+        {
+            // This will fail
```
Comment (lemunozm) [inline include/xtypes/Module.hpp]: so... should be an assert here?

File: include/xtypes/Module.hpp [hunk 4]
```diff
@@ -335,6 +350,26 @@ class Module
...
+            return static_cast<const EnumerationType<uint32_t>&>(*enumerations_32_.end()->second);
+        }
+
+        auto it = module.first->enumerations_32_.find(module.second);
+        if (it != module.first->enumerations_32_.end())
+        {
+            return static_cast<const EnumerationType<uint32_t>&>(*it->second);
+        }
+        // This will fail
```
Comment (lemunozm) [inline include/xtypes/Module.hpp]: also here?

File: include/xtypes/Module.hpp [hunk 5]
```diff
@@ -335,6 +350,26 @@ class Module
...
 
+    const EnumerationType<uint32_t>& enum_32(
+            const std::string& name) const
+    {
+        // Solve scope
+        PairModuleSymbol module = resolve_scope(name);
+        if (module.first == nullptr)
+        {
+            // This will fail
+            return static_cast<const EnumerationType<uint32_t>&>(*enumerations_32_.end()->second);
```
Comment (richiware) [inline include/xtypes/Module.hpp]: Use `xtypes_assert`.

File: include/xtypes/Module.hpp [hunk 6]
```diff
@@ -335,6 +350,26 @@ class Module
...
+        }
+
+        auto it = module.first->enumerations_32_.find(module.second);
+        if (it != module.first->enumerations_32_.end())
+        {
+            return static_cast<const EnumerationType<uint32_t>&>(*it->second);
+        }
+        // This will fail
+        return static_cast<const EnumerationType<uint32_t>&>(*enumerations_32_.end()->second);
```
Comment (richiware) [inline include/xtypes/Module.hpp]: Use `xtypes_assert`.

**Review by richiware — CHANGES_REQUESTED** [review]

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/xtypes-manual-linux/53/)

**Review by lemunozm — CHANGES_REQUESTED** [review]

File: include/xtypes/Module.hpp (hunk 2 above)
Comment (LuisGP) [reply include/xtypes/Module.hpp:282]: Yes... and not. This parameter isn't for constants created from an enumeration, but for "tricky" constants automatically created with the enumeration values by the parser.

For example, in the following IDL:
```idl
enum MyEnum
{
    A,
    B
};

struct MyStruct
{
    uint32 int_array[B];
};
```
The symbol `B` must be accessible and have the correct value in order to create the array as expected. To do so, the parser creates two constants in the current Module, `MyEnum::B` and `B` (the same for `A`, obviously). These constants aren't created by the user, and the generator must ignore them.

File: include/xtypes/Module.hpp (hunk 3 above)
Comment (LuisGP) [reply include/xtypes/Module.hpp]: Please, see the latest version of the file :).

File: include/xtypes/Module.hpp (hunk 4 above)
Comment (LuisGP) [reply include/xtypes/Module.hpp]: Again, update the branch.

**Review by richiware — APPROVED** [review]

**Pull request merged**